Libplanet/Blockchain/Policies/BlockPolicy.cs

```diff
@@ -66,134 +64,141 @@ public BlockPolicy(TimeSpan blockInterval)
-            DateTimeOffset? prevTimestamp = null;
-            IEnumerable<(long i, DifficultyExpectation)> indexedDifficulties =
-                ExpectDifficulties(blocks).Select(
-                    (exp, i) => { return ((long)i, exp); }
-                ).ToArray();
-
-            foreach (var (i, exp) in indexedDifficulties)
-            {
-                Trace.Assert(exp.Block != null);
-                Block<T> block = exp.Block;
-
-                if (i != block.Index)
-                {
-                    return new InvalidBlockIndexException(
-                        $"the expected block index is {i}, but its index is" +
-                        $" {block.Index}'"
-                    );
-                }
-
-                if (block.Difficulty < exp.Difficulty)
-                {
-                    return new InvalidBlockDifficultyException(
-                        $"the expected difficulty of the block #{i} " +
-                        $"is {exp.Difficulty}, but its difficulty is " +
-                        $"{block.Difficulty}'"
-                    );
-                }
+            Block<T> secondLastBlock = null;
+            Block<T> lastBlock = null;
 
-                if (block.PreviousHash != prevHash)
-                {
-                    if (prevHash == null)
-                    {
-                        return new InvalidBlockPreviousHashException(
-                            "the genesis block must have not previous block"
-                        );
-                    }
-
-                    return new InvalidBlockPreviousHashException(
-                        $"the block #{i} is not continuous from the " +
-                        $"block #{i - 1}; while previous block's hash is " +
-                        $"{prevHash}, the block #{i}'s pointer to " +
-                        "the previous hash refers to " +
-                        (block.PreviousHash?.ToString() ?? "nothing")
-                    );
-                }
+            for (var i = 0; i < blocks.Count; i++)
+            {
+                Block<T> block = blocks[i];
+                InvalidBlockException exception = ValidateBlockToAppend(
+                    i, secondLastBlock, lastBlock, block);
 
-                if (block.Timestamp < prevTimestamp)
+                if (!(exception is null))
                 {
-                    return new InvalidBlockTimestampException(
-                        $"the block #{i}'s timestamp ({block.Timestamp}) is " +
-                        $"earlier than the block #{i - 1}'s ({prevTimestamp})"
-                    );
+                    return exception;
                 }
 
-                prevHash = block.Hash;
-                prevTimestamp = block.Timestamp;
+                secondLastBlock = lastBlock;
+                lastBlock = block;
             }
 
             return null;
         }
 
+        /// <inheritdoc/>
+        public InvalidBlockException ValidateBlockToAppend(
```

Comment:
Do we need to expose the validation interface used when adding blocks? if `BlockChain<T>` knows detail that uses the last block instead of a `BlockPolicy`, it also seems to be a problem that can be solved using the existing `IBlockPolicy.ValidateBlocks()`.

Reply:
But again, depending on the implementation, how many previous blocks we see may change. 🤔

Reply:
Do you mean to put `ValidateBlockToAppend()` as a private method in `BlockChain<T>` instead of `IBlockPolicy`?

Reply:
yes, exactly, my first thought was removing `ValidateBlockToAppend()` and moving these logics to `BlockChain<T>.Append()`.

Reply:
To avoid problems when implementing similar interfaces, I removed `ValidateBlocks()` from` IBlockPolicy<T>` and created it as an extension method.